[WIP] File restrictions: by file type, size, number of files

arturi opened the pull request:
This gist:

```js
Uppy({
  debug: true,
  autoProceed: false,
  maxFileSize: 300000,
  maxNumberOfFiles: 1,
  allowedFileTypes: ['image', 'video']
})
```

<img width="743" alt="screen shot 2017-06-16 at 7 52 11 pm" src="https://user-images.githubusercontent.com/1199054/27248202-5dfa3b44-52ce-11e7-9f83-12cdc3514c70.png">
<img width="744" alt="screen shot 2017-06-16 at 8 00 49 pm" src="https://user-images.githubusercontent.com/1199054/27248215-838a9494-52ce-11e7-95e3-0172a2e2d868.png">

- [x] add minNumberOfFiles, which possibly means moving the `checkRestrictions` check to before upload, or checking twice
- [ ] add tests
- [x] add wildcard support or something to allow all images and PDF, mime-type of which starts with application/
- [x] add text about active restrictions to the Dashboard (#164)

arturi commented:
From #86:

> I agree that rejecting certain files would be useful from a UX standpoint rather than security. Maybe we can look up some examples of what would be a clean and powerful abstraction for that. I can imagine we'd offer power users a callback where they can supply their own function that gets passed a file and the can return false based on any logic they implement there - and for novices passing it an object with size and type properties would do some basic rejection for them.

kvz commented:
Just to be a PITA, 

What if we required 1 image and 1 audio so we could generate a video clip from that?

Thinking if something like this would make sense:

```
restrictions: {
 type: [minCount, maxCount, minSize, maxSize],
}
```

```
restrictions: {
 '*': [0, 999, 0, 999],
}
```

```
restrictions: {
 'image': [1, 1, 0, 999],
 'audio': [1, 1, 0, 999],
}
```

```
restrictions: function (files) {
   // if files[0].size > 999 return false
}
```

I realize tho this is still far from perfect and likely someone smarter than me will either just destroy this or make it less shitty :)

goto-bus-stop commented:
> add minNumberOfFiles, which possibly means moving the checkRestrictions check to before upload, or checking twice

Checking twice sounds good to me; check each time a new file is added and display any errors, and check when an upload is about to kick off. minNumberOfFiles should not prevent files from being added though of course while maxNumberOfFiles should…May need two types of checks? Or a boolean parameter to ignore the minNumberOfFiles check and others that should not prevent a new file being added?

>  add wildcard support or something to allow all images and PDF, mime-type of which starts with application/

So you could set `allowedFileTypes: ['image/*', 'application/pdf']`? I really like that!

> What if we required 1 image and 1 audio so we could generate a video clip from that?

I think this may be best addressed by accepting a function to do the check as in your last example

arturi commented:
Added the `minNumberOfFiles` check and fixed an issue discovered during live demo. Thinking about Kevin’s proposal. It does make setting options more complicated but flexibility is tempting. But if it’s indeed a rare case, maybe

> I think this may be best addressed by accepting a function to do the check

is the way to go.

arturi commented:
Current version:

```js
Uppy({
  debug: true,
  autoProceed: false,
  restrictions: {
    maxFileSize: 300000,
    maxNumberOfFiles: 5,
    minNumberOfFiles: 2,
    allowedFileTypes: ['image/*', 'video/*']
  },
  onBeforeFileAdded: (currentFile, files) => {
    if (currentFile.name === 'pitercss-IMG_0616.jpg') {
      return Promise.resolve()
    }
    return Promise.reject('this is not the file I was looking for')
  },
  onBeforeUpload: (files) => {
    if (Object.keys(files).length < 2) {
      return Promise.reject('too few files')
    }
    return Promise.resolve()
  }
})
```
